main/src/library/Nel.flix

```diff
@@ -349,6 +358,29 @@ namespace Nel {
         case Nel(x, xs) => f(x, lazy List.foldRightLazy(f, s, xs))
     }
 
+    ///
+    /// Left-associative reduction of a structure.
+    /// Applies `g` to the initial element of `l` and combines it
+    /// with the remainder of `l` using `f` going from left to right.
+    ///
+    pub def reduceLeftTo(f: (b, a) -> b & ef1, g: a -> b & ef2, l: Nel[a]): b & (ef1 and ef2) = match l {
+        case Nel(x, xs) => List.foldLeft(f, g(x), xs)
+    }
+
+    ///
+    /// Right-associative reduction of a structure.
+    /// Applies `g` to the initial element of `l` and combines it
+    /// with the remainder of `l` using `f` going from right to left.
+    ///
+    pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, l: Nel[a]): b & (ef1 and ef2) =
+        def loop(ll) = match ll { // TODO: Tail recursive impl
+            case x :: Nil => g(x)
+            case x :: xs  => f(x, loop(xs))
+            case _        => unreachable!()
+        };
+        let Nel(x, xs) = l;
+        loop(x :: xs)
```

Comment:
A big question is: When do you expect to see the effect from g? First or last? I guess you expect it first and that is also what your code does. But indeed. This function is tricky to implement. It seems you cannot just call foldRight. You have to do something like what you are doing.